src/Stratis.Bitcoin.Features.Miner.Tests/PosMintingTest.cs

```diff
@@ -351,6 +352,100 @@ public void GetNetworkWeight_UsesLast73Blocks_CalculatedNetworkWeightUsingLatest
+        {
+            bool suitable = this.WasUtxoSelectedForStaking(Network.StratisTest, 450000, 450000 - 18);
+            Assert.True(suitable);
+
+            bool suitable2 = this.WasUtxoSelectedForStaking(Network.StratisTest, 450000, 450000 - 17);
+            Assert.False(suitable2);
+        }
+
+        [Fact]
+        public void CoinstakeAge_AtTheActivation_Testnet()
+        {
+            bool suitable1 = this.WasUtxoSelectedForStaking(Network.StratisTest, 436000 - 2, 436000 - 10); // mining block before activation
+            Assert.True(suitable1);
+
+            bool suitable2 = this.WasUtxoSelectedForStaking(Network.StratisTest, 436000 - 1, 436000 - 19); // mining activation block
+            Assert.True(suitable2);
+
+            bool suitable3 = this.WasUtxoSelectedForStaking(Network.StratisTest, 436000 - 1, 436000 - 18); // mining activation block
+            Assert.False(suitable3);
+        }
+        
+        [Fact]
+        public void CoinstakeAge_BeforeActivation_Mainnet()
+        {
+            bool suitable = this.WasUtxoSelectedForStaking(Network.StratisMain, 1000, 1000 - 48); // utxo depth is 49, mining block at 50
+            Assert.True(suitable);
+
+            bool suitable2 = this.WasUtxoSelectedForStaking(Network.StratisMain, 1000, 1000 - 47); // utxo depth is 48, mining block at 49
+            Assert.False(suitable2);
+        }
+
+        [Fact]
+        public void CoinstakeAge_AfterActivation_Mainnet()
+        {
+            bool suitable = this.WasUtxoSelectedForStaking(Network.StratisMain, 950000, 950000 - 498);
+            Assert.True(suitable);
+
+            bool suitable2 = this.WasUtxoSelectedForStaking(Network.StratisMain, 950000, 950000 - 497);
+            Assert.False(suitable2);
+        }
+
+        [Fact]
+        public void CoinstakeAge_AtTheActivation_Mainnet()
+        {
+            bool suitable1 = this.WasUtxoSelectedForStaking(Network.StratisMain, 940000 - 2, 940000 - 50); // mining block before activation
+            Assert.True(suitable1);
+
+            bool suitable2 = this.WasUtxoSelectedForStaking(Network.StratisMain, 940000 - 1, 940000 - 499); // mining activation block
+            Assert.True(suitable2);
+
+            bool suitable3 = this.WasUtxoSelectedForStaking(Network.StratisMain, 940000 - 1, 940000 - 498); // mining activation block
+            Assert.False(suitable3);
+        }
+
+        private bool WasUtxoSelectedForStaking(Network network, int chainTipHeight, int utxoHeight)
+        {
+            this.network = network;
+            this.network.Consensus.Options = new PosConsensusOptions();
+            this.chain = GenerateChainWithBlockTimeAndHeight(2, this.network, 60, 0x1df88f6f);
+
+            PosMinting miner = this.InitializePosMinting();
+
+            ChainedHeader chainTip = this.chain.Tip;
+            ReflectionExtensions.SetPrivatePropertyValue(chainTip, "Height", chainTipHeight);
```

Comment:
can this be called as a proper extension method or not?
chainTip.SetPrivatePropertyValue("Height", `chainTipHeight`);

Also. despite the fact I mentioned using reflection before when you needed to do something in a test that you didn't want in the proper code. I would say it is still a code smell and suggests an flaw in the design of the component (code smell implying, it is not always necessarily bad, it just needs considering....is this a bad design when this is needed?)

Reply:
>can this be called as a proper extension method or not?
>chainTip.SetPrivatePropertyValue("Height", chainTipHeight);

Fair point. Converted `SetPrivatePropertyValue` to a proper extension method. 

>despite the fact I mentioned using reflection before when you needed to do something in a test >that you didn't want in the proper code. I would say it is still a code smell and suggests an flaw >in the design of the component (code smell implying, it is not always necessarily bad, it just >needs considering....is this a bad design when this is needed?)

I doubt that in order to write a test it's a good idea to change the design of a chained header. 

There is a way to do it alternatively- generate a concurrent chain with 950 000 headers. But do we really want to generate such a chain every test just to avoid reflection?